File: apps/backend/src/services/rag/embeddings.ts
import { chunkMarkdown } from "@yomi/shared"

export const EMBEDDING_DIMENSIONS = 1536
export const DEFAULT_EMBEDDING_MODEL = "text-embedding-3-small"
const CHUNK_CHARS = 1800
const CHUNK_OVERLAP = 220

export function chunkText(content: string): string[] {
  return chunkMarkdown(content, { targetChars: CHUNK_CHARS, overlap: CHUNK_OVERLAP })
}

export async function embedText(input: string): Promise<number[]> {
  if (!input.trim()) return []
  const apiKey = process.env["OPENAI_API_KEY"]
  if (!apiKey) throw new Error("OPENAI_API_KEY is required for Cloud RAG embeddings")
  const baseUrl = (process.env["OPENAI_BASE_URL"] ?? "https://api.aicredits.in/v1").replace(
    /\/+$/,
    "",
  )
  const model = process.env["OPENAI_EMBEDDING_MODEL"] ?? DEFAULT_EMBEDDING_MODEL
  const res = await fetch(`${baseUrl}/embeddings`, {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}`, "Content-Type": "application/json" },
    body: JSON.stringify({ model, input }),
  })
  if (!res.ok) throw new Error(`OpenAI embeddings failed: ${res.status}`)
  const body = (await res.json()) as { data?: { embedding?: number[] }[] }
  const embedding = body.data?.[0]?.embedding
  if (!Array.isArray(embedding) || embedding.length !== EMBEDDING_DIMENSIONS) {
    throw new Error(`OpenAI embedding dimensions must be ${EMBEDDING_DIMENSIONS}`)
  }
  return embedding
}

File: apps/backend/src/yomi/services/rag/embeddings.py
"""Text chunking + OpenAI embeddings for Cloud RAG.

Port of apps/backend/src/services/rag/embeddings.ts.
"""

from __future__ import annotations

import httpx

from yomi.conf import settings
from yomi.shared.chunk import ChunkOptions, chunk_markdown

EMBEDDING_DIMENSIONS = 1536
DEFAULT_EMBEDDING_MODEL = "text-embedding-3-small"
CHUNK_CHARS = 1800
CHUNK_OVERLAP = 220

_embedding_timeout = httpx.Timeout(60.0)


def chunk_text(content: str) -> list[str]:
    return chunk_markdown(content, ChunkOptions(target_chars=CHUNK_CHARS, overlap=CHUNK_OVERLAP))


async def embed_text(input_text: str) -> list[float]:
    if not input_text.strip():
        return []
    api_key = settings.openai_api_key
    if not api_key:
        raise RuntimeError("OPENAI_API_KEY is required for Cloud RAG embeddings")
    base_url = settings.openai_base_url.rstrip("/")
    model = settings.openai_embedding_model or DEFAULT_EMBEDDING_MODEL

    async with httpx.AsyncClient(timeout=_embedding_timeout) as client:
        res = await client.post(
            f"{base_url}/embeddings",
            headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
            json={"model": model, "input": input_text},
        )
    if res.status_code != 200:
        raise RuntimeError(f"OpenAI embeddings failed: {res.status_code}")
    body = res.json()
    embedding = body.get("data", [{}])[0].get("embedding")
    if not isinstance(embedding, list) or len(embedding) != EMBEDDING_DIMENSIONS:
        raise RuntimeError(f"OpenAI embedding dimensions must be {EMBEDDING_DIMENSIONS}")
    return [float(v) for v in embedding]